cluster/distributedtask/scheduler.go

```diff
@@ -0,0 +1,365 @@
+			Name: "weaviate_distributed_tasks_running",
+			Help: "Number of active distributed tasks running per namespace",
+		}, []string{"namespace"}),
+
+		stopCh: make(chan struct{}),
+	}
+}
+
+func (s *Scheduler) Start(ctx context.Context) error {
+	tasksByNamespace, err := s.listTasks(ctx)
+	if err != nil {
+		return errors.Wrap(err, "list distributed tasks")
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for namespace, provider := range s.providers {
+		provider.SetCompletionRecorder(s.completionRecorder)
+
+		var (
+			tasks         = tasksByNamespace[namespace]
+			startedTasks  = s.filterStartedTasks(tasks)
+			localTaskDesc = provider.GetLocalTasks()
+		)
+		for _, taskDesc := range localTaskDesc {
+			if _, ok := startedTasks[taskDesc]; ok {
+				continue
+			}
+
+			if err := provider.CleanupTask(taskDesc); err != nil {
+				s.logger.WithFields(logrus.Fields{
+					"namespace":   namespace,
+					"taskID":      taskDesc.ID,
+					"taskVersion": taskDesc.Version,
+					"error":       err,
+				}).Error("failed to clean up local distributed task state")
+				continue
+			}
+
+			s.logger.WithFields(logrus.Fields{
+				"namespace":   namespace,
+				"taskID":      taskDesc.ID,
+				"taskVersion": taskDesc.Version,
+			}).Info("cleaned up local distributed task state")
+		}
+
+		for desc, task := range startedTasks {
+			handle, err := provider.StartTask(task)
+			if err != nil {
+				return errors.Wrapf(err, "provider %s start task %v", namespace, desc)
+			}
+
+			s.setRunningTaskHandleWithLock(namespace, desc, handle)
+			s.logger.WithFields(logrus.Fields{
+				"namespace":   namespace,
+				"taskID":      desc.ID,
+				"taskVersion": desc.Version,
+			}).Info("started distributed task execution")
+		}
+
+		s.tasksRunning.
+			WithLabelValues(namespace).
+			Set(float64(len(startedTasks)))
+	}
+
+	enterrors.GoWrapper(s.loop, s.logger)
+
+	return nil
+}
+
+func (s *Scheduler) filterStartedTasks(tasks map[TaskDescriptor]*Task) map[TaskDescriptor]*Task {
+	return filterTasks(tasks, func(task *Task) bool {
+		return task.Status == TaskStatusStarted && !task.FinishedNodes[s.localNode]
+	})
+}
+
+func filterTasks(tasks map[TaskDescriptor]*Task, predicate func(task *Task) bool) map[TaskDescriptor]*Task {
+	filtered := make(map[TaskDescriptor]*Task, len(tasks))
+	for _, task := range tasks {
+		if !predicate(task) {
+			continue
+		}
+
+		filtered[TaskDescriptor{
+			ID:      task.ID,
+			Version: task.Version,
+		}] = task
+	}
+	return filtered
+}
+
+func (s *Scheduler) loop() {
+	ticker := s.clock.NewTicker(s.tickInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.Chan():
+			s.tick()
+		case <-s.stopCh:
+			return
+		}
+	}
+}
+
+func (s *Scheduler) tick() {
+	tasksByNamespace, err := s.listTasks(context.Background())
+	if err != nil {
+		s.sampledLogger.WithSampling(func(l logrus.FieldLogger) {
+			l.WithFields(logrus.Fields{
+				"error": err,
+			}).Error("failed to list distributed tasks")
```

Comment:
I think that there's a `WithError` method there to log the error, can we use it?